lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -0,0 +1,313 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Support for Microsoft Hyper-V emulation.
+//!
+//! Windows guests and many Linux guests can interoperate with hypervisors that
+//! implement the hypervisor described in Microsoft's Hypervisor Top-Level
+//! Functional Specification (TLFS). The behavior in this module is based on
+//! version 6.0b of the TLFS, which is available on GitHub:
+//! https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/main/tlfs/Hypervisor%20Top%20Level%20Functional%20Specification%20v6.0b.pdf
+//!
+//! Microsoft also maintains a list of minimum requirements for any hypervisor
+//! that intends to implement a Hyper-V-compatible interface:
+//! https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/main/tlfs/Requirements%20for%20Implementing%20the%20Microsoft%20Hypervisor%20Interface.pdf
+
+use std::sync::Mutex;
+
+use bits::*;
+use cpuid_utils::{CpuidIdent, CpuidSet, CpuidValues};
+use hypercall::{hypercall_page_contents, MsrHypercallValue};
+use slog::info;
+
+use crate::{
+    accessors::MemAccessor,
+    common::{GuestRegion, Lifecycle, VcpuId, PAGE_SIZE},
+    enlightenment::AddCpuidError,
+    migrate::{
+        MigrateCtx, MigrateSingle, MigrateStateError, Migrator, PayloadOffer,
+        PayloadOutput,
+    },
+    msr::{MsrId, RdmsrOutcome, WrmsrOutcome},
+    vmm::SubMapping,
+};
+
+mod bits;
+mod hypercall;
+
+const TYPE_NAME: &str = "guest-hyperv-interface";
+
+#[derive(Debug, Default)]
+struct Inner {
+    /// The last value stored in the [`bits::HV_X64_MSR_GUEST_OS_ID`] MSR.
+    msr_guest_os_id_value: u64,
+
+    /// The last value stored in the [`bits::HV_X64_MSR_HYPERCALL`] MSR.
+    msr_hypercall_value: MsrHypercallValue,
+}
+
+pub struct HyperV {
+    log: slog::Logger,
+    inner: Mutex<Inner>,
+    acc_mem: MemAccessor,
+}
+
+impl HyperV {
+    /// Creates a new Hyper-V enlightenment stack.
+    pub fn new(log: &slog::Logger) -> Self {
+        let acc_mem = MemAccessor::new_orphan();
+        let log = log.new(slog::o!("component" => "hyperv"));
+        Self { log, inner: Mutex::new(Inner::default()), acc_mem }
+    }
+
+    /// Handles a write to the HV_X64_MSR_GUEST_OS_ID register.
+    fn handle_wrmsr_guest_os_id(&self, value: u64) -> WrmsrOutcome {
+        let mut inner = self.inner.lock().unwrap();
+
+        // TLFS section 3.13 says that the hypercall page "becomes disabled" if
+        // the guest OS ID register is cleared after the hypercall register is
+        // set. It also specifies that attempts to set the Enabled bit in that
+        // register will be ignored if the guest OS ID is zeroed, so handle this
+        // case by clearing the hypercall MSR's Enabled bit but otherwise
+        // leaving the hypercall page untouched (as would happen if the guest
+        // manually cleared this bit).
+        if value == 0 {
+            info!(&self.log, "guest cleared HV_X64_MSR_GUEST_OS_ID");
```

Comment:
Do we really want this at `info` level?

Reply:
I'd like to be able to walk up to a system where enlightenments don't seem to be working properly and find out what state it's in, but I think we can do that without allowing the guest to spam things into the logs. I've at least switched this to a USDT probe in 5bf277f7; this would also be a good thing to include in #482.